drag_and_drop_v2/public/js/drag_and_drop.js

```diff
@@ -128,7 +133,28 @@ function DragAndDropBlock(runtime, element) {
         });
     };
 
-    var submitLocation = function(item_id, zone, top, left) {
+    var submitLocation = function(state, item_id, zone, top, left) {
+        // If height of items wrapper has changed, adjust vertical location of item
+        var newItemsWrapperHeight = $('.items', element).height();
+
+        if (newItemsWrapperHeight < itemsWrapperHeight) {
+            var heightDelta = itemsWrapperHeight - newItemsWrapperHeight,
+                item = state.state.items[item_id],
+                newTop = parseFloat(item.top) - heightDelta,
+                itemElement = $('.option[data-value="' + item_id + '"]', element);
+
+            // Update top
+            top = newTop + 'px';
+            // Update state
+            item.top = top;
+            state.state.items[item_id] = item;
+            // Update position
+            itemElement.css('top', top);
```

Comment:
+1. I like the first idea above (move it to the .target element). In fact that will be almost necessary for my task, where I make the whole Xblock proportionally sized and resizable.

I think it's fine if the `.items` shrinks as you pull items out from it (that will actually improve usability if there are dozens of items, since the user can do less scrolling as the items stay close to the `.target` rather than having blank spaces appear there). However, the `items` should not shrink before or during the drag-and-drop. If it does shrink, it should shrink at the end, after the user has dropped an item. A simple way to achieve this is to set `$items.height($items.height())` in ondragstart and `$items.height('auto')` in ondragend. (Not sure if there's a more virtual-domy way to achieve that, but I know the jQuery code is already doing all kinds of direct DOM manipulation like that during the drag process anyways.)